scipy/stats/_multivariate.py

```diff
@@ -5067,3 +5068,257 @@ def rvs(self, size=1, random_state=None):
+
+        ntot_row = np.sum(row)
+        ntot_col = np.sum(col)
+        if ntot_row != ntot_col:
+            raise ValueError("Sums over all rows and columns must be equal.")
+
+        return r, c, ntot_row
+
+    def mean(self, row, col):
+        """Mean of distribution of conditional tables.
+        %(_doc_mean_callparams)s
+        Returns
+        -------
+        mean: ndarray
+            The mean of the distribution
+
+        Notes
+        -----
+        %(_doc_callparams_note)s
+        """
+        r, c, n = self._process_parameters(row, col)
+        return np.outer(r, c) / n
+
+    def rvs(self, row, col, size=1, method=None, random_state=None):
+        """Draw random tables with fixed column and row marginals.
+
+        Parameters
+        ----------%(_doc_callparams)s
+        size : integer, optional
+            Number of samples to draw (default 1).
+        method : str, optional
+            Which method to use, "shuffle" or "patefield". If None (default),
+            an attempt is made to select the fastest method for this input.
+        %(_doc_random_state)s
+
+        Returns
+        -------
+        rvs : ndarray
+            Random 2D tables of shape (`size`, `len(row)`, `len(col)`)
+
+        Notes
+        -----
+        %(_doc_callparams_note)s
+        """
+        r, c, n = self._process_parameters(row, col)
+        random_state = self._get_random_state(random_state)
+        meth = self._process_method(method, r, c, n)
+        return meth(r, c, n, size, random_state)
+
+    @classmethod
+    def _process_method(cls, method, r, c, n):
+        known_methods = {
+            None: cls._rvs_select(r, c, n),
+            "shuffle": cls._rvs_shuffle,
+            "patefield": cls._rvs_patefield,
+        }
+        try:
+            return known_methods[method]
+        except KeyError:
+            raise ValueError(f"{method} not recognized, must be one of {{{', '.join(known_methods.keys())}}}")
+
+    @classmethod
+    def _rvs_select(cls, r, c, n):
+        # TODO find optimal empirical threshold with benchmarks,
+        # for now we always return "shuffle", because "patefield"
+        # is not yet implemented. Example:
+        # if n > 1000:
+        #     return "patefield"
+        return cls._rvs_shuffle
+
+    @staticmethod
+    def _rvs_shuffle(row, col, tot, size, random_state):
+        # could be cached in the frozen distribution
+        k = 0
+        col_idx = np.empty(tot, dtype=np.int_)
+        for i, ci in enumerate(col):
+            col_idx[k:k+ci] = i
+            k += ci
+
+        matrix = np.zeros((size, len(row), len(col)))
+        for o in range(size):
+            random_state.shuffle(col_idx)
+            k = 0
+            # slow in pure Python, will be replaced by C code
+            for i, ri in enumerate(row):
+                for j in range(ri):
+                    matrix[o, i, col_idx[k+j]] += 1
+                k += ri
+        return matrix
```

Comment:
Would this be OK, so I don't have to think about the algorithm yet?
```suggestion
        x = np.concatenate([[i]*j for i, j in enumerate(row)])
        y = np.concatenate([[i]*j for i, j in enumerate(col)])

        def crosstab(x, y):
            return contingency.crosstab(x, y).count

        tables = [crosstab(x, random_state.permutation(y))
                  for i in range(size)]
        return np.asarray(tables)
```
Not that it matters, but it seems to be faster for larger tables.
You'd need to change line 17 to `from . import _mvn, _covariance, contingency`.

Reply:
Yes, that is better. I will use np.repeat instead of np.concatenate, though.